### ISSUE TITLE
Investigate How to Call MPI Directly

### ISSUE THREAD

**Issue by bosawt** [department-of-veterans-affairs/va.gov-team#19975]

**Comment by bosawt** [comment]
I think we can't actually do this yet, so I'm changing the name of the title from `Create example mock entry in MPI for local development to use with relationship associations`, to `Investigate how to call MPI directly`. The reason we can't do this yet, is making a call to MPI and requesting relationship information currently doesn't work, even if the user exists and the relationship information is blank. So we need to wait until we can at least get back blank information before we can mock a user with this information

**Comment by bosawt** [comment]
Instead, here is information on how to call MPI directly:

# MPI Service Testing
#oddball/reference/mpi
#oddball/reference/mvi
#oddball/testing/mpi
#oddball/testing/mvi

### How to Connect
* To connect to MPI Service
	* Need a certificate
	* Need to be on the VA network (or communicate with a proxy)
* It can be difficult to connect to the service locally for this reason, so one of the easier ways is to find a PR with a deployment built with Jenkins
	* Look for verbiage that looks like this:
[image:81BC834D-E835-42A6-BED1-4427DFE0A8D0-4237-00017D1AE81BDD1C/Screen Shot 2021-02-18 at 3.59.54 PM.png]
`msg: Review instance available at http://3e331238d967c1237fd87ee8719e0b12.review.vetsgov-internal/, or via SSH at ip-172-30-19-182.us-gov-west-1.compute.internal`
* You can simply `ssh ip-172-30-19-182.us-gov-west-1.compute.internal` to connect to the deployed instance
* You can find the relevant path with `cat config/settings.local.yml`, grep for `mvi`, or `IdMWebService` to find the path
* You can then `cURL` the path, although make sure to add the `-k` setting on the `cURL`, it authenticates with a self signed SSL, so `cURL` is not happy unless you add the setting to ignore this fact

### WSDL
* Request to retrieve WSDL MPI Schema:
``` bash
curl -k https://internal-dsva-vagov-staging-fwdproxy-1821450725.us-gov-west-1.elb.amazonaws.com:4434/psim_webservice/stage1a/IdMWebService?WSDL
```
* Response from the above call:
``` xml
<?xml version="1.0" encoding="UTF-8"?>

<wsdl:definitions xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/" xmlns:soap11="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:soap12="http://schemas.xmlsoap.org/wsdl/soap12/" xmlns:http="http://schemas.xmlsoap.org/wsdl/http/" xmlns:mime="http://schemas.xmlsoap.org/wsdl/mime/" xmlns:wsp="http://www.w3.org/ns/ws-policy" xmlns:wsp200409="http://schemas.xmlsoap.org/ws/2004/09/policy" xmlns:wsp200607="http://www.w3.org/2006/07/ws-policy" xmlns:ns0="http://vaww.oed.oit.va.gov" targetNamespace="http://vaww.oed.oit.va.gov">
<wsdl:types xmlns:xsd="http://www.w3.org/2001/XMLSchema">
<xsd:schema>
<xsd:import schemaLocation="IdMWebService.xsd1.xsd" namespace="http://vaww.oed.oit.va.gov"/></xsd:schema></wsdl:types>
<wsdl:message name="MCCI_IN000002UV01">
<wsdl:part name="returnXML" element="xsns:MCCI_IN000002UV01" xmlns:xsns="http://vaww.oed.oit.va.gov"/></wsdl:message>
<wsdl:message name="PRPA_IN201301UV02">
<wsdl:part name="inboundXML" element="xsns:PRPA_IN201301UV02" xmlns:xsns="http://vaww.oed.oit.va.gov"/></wsdl:message>
<wsdl:message name="PRPA_IN201302UV02">
<wsdl:part name="inboundXML" element="xsns:PRPA_IN201302UV02" xmlns:xsns="http://vaww.oed.oit.va.gov"/></wsdl:message>
<wsdl:message name="PRPA_IN201304UV02">
<wsdl:part name="inboundXML" element="xsns:PRPA_IN201304UV02" xmlns:xsns="http://vaww.oed.oit.va.gov"/></wsdl:message>
<wsdl:message name="PRPA_IN201305UV02">
<wsdl:part name="inboundXML" element="xsns:PRPA_IN201305UV02" xmlns:xsns="http://vaww.oed.oit.va.gov"/></wsdl:message>
<wsdl:message name="PRPA_IN201306UV02">
<wsdl:part name="returnXML" element="xsns:PRPA_IN201306UV02" xmlns:xsns="http://vaww.oed.oit.va.gov"/></wsdl:message>
<wsdl:message name="PRPA_IN201309UV02">
<wsdl:part name="inboundXML" element="xsns:PRPA_IN201309UV02" xmlns:xsns="http://vaww.oed.oit.va.gov"/></wsdl:message>
<wsdl:message name="PRPA_IN201310UV02">
<wsdl:part name="returnXML" element="xsns:PRPA_IN201310UV02" xmlns:xsns="http://vaww.oed.oit.va.gov"/></wsdl:message>
<wsdl:portType name="VAIdMPort">
<wsdl:operation name="PRPA_IN201301UV02">
<wsdl:input name="PRPA_IN201301UV02Request" message="ns0:PRPA_IN201301UV02"/>
<wsdl:output name="PRPA_IN201301UV02Response" message="ns0:MCCI_IN000002UV01"/></wsdl:operation>
<wsdl:operation name="PRPA_IN201302UV02">
<wsdl:input name="PRPA_IN201302UV02Request" message="ns0:PRPA_IN201302UV02"/>
<wsdl:output name="PRPA_IN201302UV02Response" message="ns0:MCCI_IN000002UV01"/></wsdl:operation>
<wsdl:operation name="PRPA_IN201304UV02">
<wsdl:input name="PRPA_IN201304UV02Request" message="ns0:PRPA_IN201304UV02"/>
<wsdl:output name="PRPA_IN201304UV02Response" message="ns0:MCCI_IN000002UV01"/></wsdl:operation>
<wsdl:operation name="PRPA_IN201305UV02">
<wsdl:input name="PRPA_IN201305UV02Request" message="ns0:PRPA_IN201305UV02"/>
<wsdl:output name="PRPA_IN201305UV02Response" message="ns0:PRPA_IN201306UV02"/></wsdl:operation>
<wsdl:operation name="PRPA_IN201309UV02">
<wsdl:input name="PRPA_IN201309UV02Request" message="ns0:PRPA_IN201309UV02"/>
<wsdl:output name="PRPA_IN201309UV02Response" message="ns0:PRPA_IN201310UV02"/></wsdl:operation></wsdl:portType>
<wsdl:binding name="VAIdMPort" type="ns0:VAIdMPort">
<soap11:binding transport="http://schemas.xmlsoap.org/soap/http" style="document"/>
<wsdl:operation name="PRPA_IN201301UV02">
<soap11:operation soapAction="" style="document"/>
<wsdl:input name="PRPA_IN201301UV02Request">
<soap11:body use="literal"/></wsdl:input>
<wsdl:output name="PRPA_IN201301UV02Response">
<soap11:body use="literal"/></wsdl:output></wsdl:operation>
<wsdl:operation name="PRPA_IN201302UV02">
<soap11:operation soapAction="" style="document"/>
<wsdl:input name="PRPA_IN201302UV02Request">
<soap11:body use="literal"/></wsdl:input>
<wsdl:output name="PRPA_IN201302UV02Response">
<soap11:body use="literal"/></wsdl:output></wsdl:operation>
<wsdl:operation name="PRPA_IN201304UV02">
<soap11:operation soapAction="" style="document"/>
<wsdl:input name="PRPA_IN201304UV02Request">
<soap11:body use="literal"/></wsdl:input>
<wsdl:output name="PRPA_IN201304UV02Response">
<soap11:body use="literal"/></wsdl:output></wsdl:operation>
<wsdl:operation name="PRPA_IN201305UV02">
<soap11:operation soapAction="" style="document"/>
<wsdl:input name="PRPA_IN201305UV02Request">
<soap11:body use="literal"/></wsdl:input>
<wsdl:output name="PRPA_IN201305UV02Response">
<soap11:body use="literal"/></wsdl:output></wsdl:operation>
<wsdl:operation name="PRPA_IN201309UV02">
<soap11:operation soapAction="" style="document"/>
<wsdl:input name="PRPA_IN201309UV02Request">
<soap11:body use="literal"/></wsdl:input>
<wsdl:output name="PRPA_IN201309UV02Response">
<soap11:body use="literal"/></wsdl:output></wsdl:operation></wsdl:binding>
<wsdl:service name="VAIdM">
<wsdl:port name="VAIdMPort" binding="ns0:VAIdMPort">
<soap11:address location="https://sqa.services.eauth.va.gov:9303/psim_webservice/stage1a/IdMWebService"/></wsdl:port>
<wsdl:port name="VAIdMPort.0" binding="ns0:VAIdMPort">
<soap11:address location="https://sqa.services.eauth.va.gov:9303/psim_webservice/stage1a/IdMWebService"/></wsdl:port>
<wsdl:port name="VAIdMPort.1" binding="ns0:VAIdMPort">
<soap11:address location="https://sqa.services.eauth.va.gov:9303/psim_webservice/stage1a/IdMWebService"/></wsdl:port></wsdl:service></wsdl:definitions>
```
### Find Person
* Request for Find Person call:
``` bash
curl -v -k -X POST -H 'Accept: text/xml;charset=UTF-8' -H 'Content-Type: text/xml;charset=UTF-8' -H 'User-Agent: Vets.gov Agent' -H 'soapaction: PRPA_IN201305UV02' -H 'Date: Thu, 18 Feb 2021 17:57:14 GMT' -H 'Content-Length: 1884' -d '
<env:Envelope xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:env="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <env:Header/>
  <env:Body>
    <idm:PRPA_IN201305UV02 xmlns:idm="http://vaww.oed.oit.va.gov" xmlns:xsi="http://www.w3.org/2001/XMLSchema‐instance" xsi:schemaLocation="urn:hl7‐org:v3 ../../schema/HL7V3/NE2008/multicacheschemas/PRPA_IN201305UV02.xsd" xmlns="urn:hl7‐org:v3" ITSVersion="XML_1.0">
      <id root="1.2.840.114350.1.13.0.1.7.1.1" extension="200VGOV-ba391f62-69c6-4b71-bfb3-d5fe2877b805"/>
      <creationTime value="20210218175714"/>
      <versionCode code="4.1"/>
      <interactionId root="2.16.840.1.113883.1.6" extension="PRPA_IN201305UV02"/>
      <processingCode code="T"/>
      <processingModeCode code="T"/>
      <acceptAckCode code="AL"/>
      <receiver typeCode="RCV">
        <device classCode="DEV" determinerCode="INSTANCE">
          <id root="1.2.840.114350.1.13.999.234" extension="200M"/>
        </device>
      </receiver>
      <sender typeCode="SND">
        <device classCode="DEV" determinerCode="INSTANCE">
          <id root="2.16.840.1.113883.4.349" extension="200VGOV"/>
        </device>
      </sender>
      <controlActProcess classCode="CACT" moodCode="EVN">
        <code code="PRPA_TE201305UV02" codeSystem="2.16.840.1.113883.1.6"/>
        <queryByParameter>
          <queryId root="1.2.840.114350.1.13.28.1.18.5.999" extension="18204"/>
          <statusCode code="new"/>
          <modifyCode code="MVI.COMP2"/>
          <initialQuantity value="1"/>
          <parameterList>
            <id root="2.16.840.1.113883.4.349" extension="1013174892V530328"/>
          </parameterList>
        </queryByParameter>
      </controlActProcess>
    </idm:PRPA_IN201305UV02>
  </env:Body>
</env:Envelope>
' "https://internal-dsva-vagov-staging-fwdproxy-1821450725.us-gov-west-1.elb.amazonaws.com:4434/psim_webservice/stage1a/IdMWebService"
```
* Response from the above request:
	* Logged part of response:
``` bash
*   Trying 10.247.34.186...
* TCP_NODELAY set
* Connected to internal-dsva-vagov-staging-fwdproxy-1821450725.us-gov-west-1.elb.amazonaws.com (10.247.34.186) port 4434 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: C=US; ST=.; L=.; O=.; CN=*.us-gov-west-1.elb.amazonaws.com
*  start date: Feb 23 23:42:25 2017 GMT
*  expire date: Feb 21 23:42:25 2027 GMT
*  issuer: C=US; ST=.; L=.; O=.; CN=*.us-gov-west-1.elb.amazonaws.com
*  SSL certificate verify result: self signed certificate (18), continuing anyway.
> POST /psim_webservice/stage1a/IdMWebService HTTP/1.1
> Host: internal-dsva-vagov-staging-fwdproxy-1821450725.us-gov-west-1.elb.amazonaws.com:4434
> Accept: text/xml;charset=UTF-8
> Content-Type: text/xml;charset=UTF-8
> User-Agent: Vets.gov Agent
> soapaction: PRPA_IN201305UV02
> Date: Thu, 18 Feb 2021 17:57:14 GMT
> Content-Length: 1884
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
< x-note: Gateway Ack
* We are completely uploaded and fine
< HTTP/1.1 200 OK
< x-backside-transport: OK OK,OK OK
< transfer-encoding: chunked
< date: Fri, 19 Feb 2021 00:07:27 GMT
< content-type: text/xml
< x-global-transaction-id: f8ba5315602f013e0a97bca2
< 
```
	* XML part of response:
``` xml
<?xml version="1.0" encoding="UTF-8"?>
<env:Envelope xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/"
  xmlns:xsd="http://www.w3.org/2001/XMLSchema"
  xmlns:env="http://schemas.xmlsoap.org/soap/envelope/"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <env:Header/>
  <env:Body>
    <idm:PRPA_IN201306UV02 ITSVersion="XML_1.0" xsi:schemaLocation="urn:hl7-org:v3 ../../schema/HL7V3/NE2008/multicacheschemas/PRPA_IN201306UV02.xsd"
      xmlns="urn:hl7-org:v3"
      xmlns:idm="http://vaww.oed.oit.va.gov">
      <id extension="WSDOC2102181916499501180113763" root="2.16.840.1.113883.4.349"/>
      <creationTime value="20210218191649"/>
      <versionCode code="4.1"/>
      <interactionId extension="PRPA_IN201306UV02" root="2.16.840.1.113883.1.6"/>
      <processingCode code="T"/>
      <processingModeCode code="T"/>
      <acceptAckCode code="NE"/>
      <receiver typeCode="RCV">
        <device determinerCode="INSTANCE" classCode="DEV">
          <id extension="200VGOV" root="2.16.840.1.113883.4.349"/>
        </device>
      </receiver>
      <sender typeCode="SND">
        <device determinerCode="INSTANCE" classCode="DEV">
          <id extension="200M" root="2.16.840.1.113883.4.349"/>
        </device>
      </sender>
      <acknowledgement>
        <typeCode code="AA"/>
        <targetMessage>
          <id extension="200VGOV-ba391f62-69c6-4b71-bfb3-d5fe2877b805" root="1.2.840.114350.1.13.0.1.7.1.1"/>
        </targetMessage>
      </acknowledgement>
      <controlActProcess classCode="CACT" moodCode="EVN">
        <code codeSystem="2.16.840.1.113883.1.6" code="PRPA_TE201306UV02"/>
        <subject typeCode="SUBJ">
          <registrationEvent classCode="REG" moodCode="EVN">
            <id nullFlavor="NA"/>
            <statusCode code="active"/>
            <subject1 typeCode="SBJ">
              <patient classCode="PAT">
                <id extension="1013174892V530328^NI^200M^USVHA^P" root="2.16.840.1.113883.4.349"/>
                <id extension="17106793^PI^200MHS^USVHA^A" root="2.16.840.1.113883.4.349"/>
                <id extension="49f4c9c79eef46ee80e5cb1e63b3942b^PN^200VIDM^USDVA^A" root="2.16.840.1.113883.4.349"/>
                <id extension="1013174892^PN^200PROV^USDVA^A" root="2.16.840.1.113883.4.349"/>
                <id extension="2107313977^NI^200DOD^USDOD^A" root="2.16.840.1.113883.3.42.10001.100001.12"/>
                <id extension="7219985^PI^983^USVHA^A" root="2.16.840.1.113883.4.349"/>
                <statusCode code="active"/>
                <patientPerson>
                  <name use="L">
                    <given>PATRYKK</given>
                    <family>PIISAWE</family>
                  </name>
                  <administrativeGenderCode code="M"/>
                  <birthTime value="19810403"/>
                  <asOtherIDs classCode="SSN">
                    <id extension="125201174" root="2.16.840.1.113883.4.1"/>
                    <scopingOrganization determinerCode="INSTANCE" classCode="ORG">
                      <id root="1.2.840.114350.1.13.99997.2.3412"/>
                    </scopingOrganization>
                  </asOtherIDs>
                </patientPerson>
                <subjectOf1>
                  <queryMatchObservation classCode="COND" moodCode="EVN">
                    <code code="IHE_PDQ"/>
                    <value nullFlavor="NA" xsi:type="INT"/>
                  </queryMatchObservation>
                </subjectOf1>
                <subjectOf2 typeCode="SBJ">
                  <administrativeObservation classCode="VERIF">
                    <code codeSystem="2.16.840.1.113883.4.349" code="PERSON_TYPE" displayName="Person Type"/>
                    <value xsi:type="CD" code="PAT" displayName="Patient"/>
                  </administrativeObservation>
                </subjectOf2>
              </patient>
            </subject1>
            <custodian typeCode="CST">
              <assignedEntity classCode="ASSIGNED">
                <id root="2.16.840.1.113883.4.349"/>
              </assignedEntity>
            </custodian>
          </registrationEvent>* Connection #0 to host internal-dsva-vagov-staging-fwdproxy-1821450725.us-gov-west-1.elb.amazonaws.com left intact
        </subject>
        <queryAck>
          <queryId extension="18204" root="1.2.840.114350.1.13.28.1.18.5.999"/>
          <queryResponseCode code="OK"/>
          <resultCurrentQuantity value="1"/>
        </queryAck>
        <queryByParameter>
          <queryId extension="18204" root="1.2.840.114350.1.13.28.1.18.5.999"/>
          <statusCode code="new"/>
          <modifyCode code="MVI.COMP2"/>
          <initialQuantity value="1"/>
          <parameterList>
            <id extension="1013174892V530328" root="2.16.840.1.113883.4.349"/>
          </parameterList>
        </queryByParameter>
      </controlActProcess>
    </idm:PRPA_IN201306UV02>
  </env:Body>
</env:Envelope>
```
